File: leetcode/CPP/34. Find First and Last Position of Element in Sorted Array.cpp
// https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
class Solution {
public:
    int getRange(vector<int>& nums, int target, int isLeft){
        int length = nums.size();
        int left = 0;
        int right = length-1;
        while(left<=right){
            int mid = (left + right)/2;
            if(target<nums[mid]) right = mid-1;
            else if(target>nums[mid]) left = mid+1;
            
            else{
                if(isLeft){
                    if(mid == 0 || nums[mid-1]!= target){
                        return mid;
                    }
                    else right = mid-1;
                }else {
                    if(mid == length-1 || nums[mid+1]!=target){
                        return mid;
                    }
                    else left = mid+1;
                }
            }
        }
        return -1;
    }
    
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> output(2,-1);
        output[0] = getRange(nums, target, 1);
        output[1] = getRange(nums, target, 0);
        return output;
    }
};

//  vector<int> searchRange(vector<int>& nums, int target) {
//     vector<int> output(2,-1);
//     if(nums.size() == 0){
//         return output;
//     }
//     auto min = lower_bound(nums.begin(), nums.end(),target);
//     if(*min != target){
//         return output;
//     }
//     auto max = upper_bound(nums.begin(), nums.end(), target);
//     if(max == nums.end()){
//         return output;
//     }
//     output[0] = min-nums.begin();
//     output[1] = max-nums.begin()-1;
//     return output;
// }

File: Leetcode/Java/34-Find-First-and-Last-Position-of-Element-in-Sorted-Array.java
// Solution By Tekane Chaitanya
// Question link :- https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] finalAns = {-1, -1};
        // find first occurence
        finalAns[0] = search(nums, target, true);
        if(finalAns[0] != -1){
            // find last occurence
            finalAns[1] = search(nums, target, false);
        }
        return finalAns;
    }
    int search(int[] nums, int target, boolean startIndex){
        int ans = -1;
        int start = 0;
        int end = nums.length - 1;
        while(start<=end){
            int mid = start + (end-start) / 2;
            if(target > nums[mid]){
                start = mid + 1;
            } else if(target < nums[mid]){
                end = mid - 1;
            } else {
                // potential answer found
                ans = mid;
                if(startIndex){
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            }
        }
        return ans;
    }
    
}

// Time Complexity :- O(log n)
